lib/trento/domain/cluster/cluster.ex

```diff
@@ -204,6 +205,48 @@ defmodule Trento.Domain.Cluster do
   def execute(%Cluster{cluster_id: nil}, _),
     do: {:error, :cluster_not_registered}
 
+  # Restoration, when a RegisterClusterHost command is received for a deregistered Cluster
+  # the cluster is restored, the host is added to cluster and if the host is a DC
+  # cluster details will be updated
+  def execute(
+        %Cluster{deregistered_at: deregistered_at, cluster_id: cluster_id},
+        %RegisterClusterHost{
+          host_id: host_id,
+          designated_controller: false
+        }
+      )
+      when not is_nil(deregistered_at) do
+    [
+      %ClusterRestored{cluster_id: cluster_id},
+      %HostAddedToCluster{
+        cluster_id: cluster_id,
+        host_id: host_id
+      }
+    ]
+  end
+
+  def execute(
+        %Cluster{deregistered_at: deregistered_at, cluster_id: cluster_id} = cluster,
+        %RegisterClusterHost{
+          host_id: host_id,
+          designated_controller: true
+        } = command
+      )
+      when not is_nil(deregistered_at) do
+    cluster
+    |> Multi.new()
+    |> Multi.execute(fn _ ->
+      %ClusterRestored{cluster_id: cluster_id}
+    end)
+    |> Multi.execute(fn _ ->
+      %HostAddedToCluster{
```

Comment:
Could this event `HostAddedToCluster` be emitted twice?
Asking because later in `maybe_update_cluster` there is also a `maybe_emit_host_added_to_cluster_event`
:thinking:

Reply:
Actually no, because the multi is involved, the maybe check if the host is already present, and if yes, does not emit event twice :D, this function is idempotent on both the flows restoration/registration

Reply:
You can see the list of emitted events in the test, when we add a dc node, and the host added to cluster event is emitted once in the entire restoring/update flow

Reply:
@CDimonaco I think the tests would pass without this explicit event as well.
I think the multi in fact is doing nothing, as the `HostRestored` simply changes the `deregistered_at` field, which doesn't affect the `maybe_update_cluster`.
Could you at least give a try?
Remove that event and see if the tests pass

Reply:
the test passes of course, I answered in your comment below 👀

Reply:
Ah-ha! Now I recall: having multi in place makes sure that the state is up to date at each step so at the first `HostAddedToCluster` we'll have the aggregate state updated so it won't be re-added in the following `maybe_update_cluster`